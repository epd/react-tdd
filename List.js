import React from 'react';

const doSomeListMagic = (item, idx) =>
    idx % 2 === 0 ? item.toUpperCase() : item.toLowerCase();

const List = ({items}) => (
    <ul>
        {items.map((item, idx) => (
            <li key={idx}>{doSomeListMagic(item, idx)}</li>
        ))}

        {/* <li key={idx}>{item.toUpperCase()}</li> */}
    </ul>
);

export default List;
