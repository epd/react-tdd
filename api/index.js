export const getOfficesList = () =>
    fetch('/api/offices-list', {
        headers: {
            'accepts': 'application/json', // prettier-ignore
            'content-type': 'application/json',
        },
    }).then((res) => res.json());
