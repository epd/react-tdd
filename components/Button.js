import React from 'react';

export const buttonClickHandler = () => window.alert('In your face.');

const Button = ({onClick, text}) => (
    <button onClick={onClick} type="submit">
        {text}
    </button>
);

export default Button;
