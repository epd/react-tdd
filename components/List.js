import React from 'react';

import {conditionalCaseTransform} from '../utils';

const List = ({items}) => (
    <ul>
        {items.map((item, idx) => (
            <li key={idx}>{conditionalCaseTransform(item, idx)}</li>
        ))}
    </ul>
);

export default List;
