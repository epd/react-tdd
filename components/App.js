import React, {useState} from 'react';

import Button from './Button';
import List from './List';

import {getOfficesList} from '../api';

const App = () => {
    const [offices, setOfficesList] = useState([]);
    const retreiveAndRenderOffices = () =>
        getOfficesList().then(setOfficesList);

    if (offices.length) {
        return <List items={offices} />;
    }

    return <Button onClick={retreiveAndRenderOffices} text="Load Offices" />;
};

export default App;
