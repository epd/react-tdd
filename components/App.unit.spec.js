import React from 'react';
import {act} from 'react-dom/test-utils';
import {mount, shallow} from 'enzyme';

import App from './App';
import Button from './Button';
import List from './List';

import {getOfficesList} from '../api';

jest.mock('../api');

describe('App', () => {
    beforeEach(() => {
        getOfficesList
            .mockReset()
            .mockImplementation(() => Promise.resolve([]));
    });

    it('renders a button by default', () => {
        const wrapper = shallow(<App />);

        expect(wrapper.find(Button).exists()).toBe(true);
        expect(wrapper.find(Button).dive().text()).toBe('Load Offices'); // prettier-ignore
    });

    it('loads list of offices on click of the button', () => {
        const wrapper = shallow(<App />);

        wrapper.find(Button).simulate('click');

        expect(getOfficesList).toHaveBeenCalled();
    });

    it('shows the list of offices after loading', async () => {
        const mockResponse = Promise.resolve([
            'Villena',
            'Madrid',
            'San Francisco',
            'Vancouver',
            'Nashville',
        ]);
        getOfficesList.mockImplementation(() => mockResponse);

        const wrapper = mount(<App />);

        expect(wrapper.find(List).exists()).toBe(false);

        await act(() => {
            wrapper.find(Button).simulate('click');
            return mockResponse;
        });

        wrapper.update();

        expect(wrapper.find(List).exists()).toBe(true);
        expect(wrapper.find(List)).toMatchSnapshot();
    });
});
