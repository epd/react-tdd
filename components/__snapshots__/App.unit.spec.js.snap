// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App shows the list of offices after loading 1`] = `
<List
  items={
    Array [
      "Villena",
      "Madrid",
      "San Francisco",
      "Vancouver",
      "Nashville",
    ]
  }
>
  <ul>
    <li
      key="0"
    >
      VILLENA
    </li>
    <li
      key="1"
    >
      madrid
    </li>
    <li
      key="2"
    >
      SAN FRANCISCO
    </li>
    <li
      key="3"
    >
      vancouver
    </li>
    <li
      key="4"
    >
      NASHVILLE
    </li>
  </ul>
</List>
`;
